### DISCUSSION TITLE
Stop loss spikes

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#176] [Archived]
<USER>>>> rui<br/> <DATE> [April 6, 2020, 7:34pm]<br/> <br/> Hi, I am training a Tacotron2 model on a proprietary dataset with<br/> around *slash* <br/> 1000 clips and a total duration of approximately 90 minutes. I've been *slash* <br/> able to achieve okay'ish results, with decent generalization to new *slash* <br/> test sentences. I am puzzled, however, by the stop loss *slash* <br/> performance. Specifically, it increases a lot once the gradual r *slash* <br/> training gets to r=2 (15k steps) and r=1 (27k steps).<br/> <br/> Any idea of what may be causing this?<br/> <br/> The config file used is as follows:<br/> <br/> > { *slash* <br/> > 'model': 'Tacotron2', // one of the model in models/ *slash* <br/> > 'run_name': 'ljspeech-stft_params', *slash* <br/> > 'run_description': 'tacotron2 cosntant stf parameters',<br/> ><br/> > // AUDIO PARAMETERS<br/> > 'audio':{<br/> > // Audio processing parameters<br/> > 'num_mels': 80, // size of the mel spec frame.<br/> > 'num_freq': 1025, // number of stft frequency levels. Size of the linear spectogram frame.<br/> > 'sample_rate': 22050, // DATASET-RELATED: wav sample-rate. If different than the original data, it is resampled.<br/> > 'win_length': 1024, // stft window length in ms.<br/> > 'hop_length': 256, // stft window hop-lengh in ms.<br/> > 'frame_length_ms': null, // stft window length in ms.If null, 'win_length' is used.<br/> > 'frame_shift_ms': null, // stft window hop-lengh in ms. If null, 'hop_length' is used.<br/> > 'preemphasis': 0.98, // pre-emphasis to reduce spec noise and make it more structured. If 0.0, no -pre-emphasis.<br/> > 'min_level_db': -100, // normalization range<br/> > 'ref_level_db': 20, // reference level db, theoretically 20db is the sound of air.<br/> > 'power': 1.5, // value to sharpen wav signals after GL algorithm.<br/> > 'griffin_lim_iters': 60,// #griffin-lim iterations. 30-60 is a good range. Larger the value, slower the generation.<br/> > // Normalization parameters<br/> > 'signal_norm': true, // normalize the spec values in range [0, 1]<br/> > 'symmetric_norm': true, // move normalization to range [-1, 1]<br/> > 'max_norm': 4.0, // scale normalization to range [-max_norm, max_norm] or [0, max_norm]<br/> > 'clip_norm': true, // clip normalized values into the range.<br/> > 'mel_fmin': 50.0, // minimum freq level for mel-spec. ~50 for male and ~95 for female voices. Tune for dataset!!<br/> > 'mel_fmax': 8000.0, // maximum freq level for mel-spec. Tune for dataset!!<br/> > 'do_trim_silence': false, // enable trimming of slience of audio as you load it. LJspeech (false), TWEB (false), Nancy (true)<br/> > 'trim_db': 60 // threshold for timming silence. Set this according to your dataset.<br/> > },<br/> ><br/> > // VOCABULARY PARAMETERS<br/> > // if custom character set is not defined,<br/> > // default set in symbols.py is used<br/> > 'characters':{<br/> > 'pad': '_',<br/> > 'eos': '~',<br/> > 'bos': '^',<br/> > 'characters': 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzÁÀÂÃÉÊÍÓÔÕÚáàâãéêíóôõúç!'(),-.:;? ',<br/> > 'punctuations':'!'(),-.:;? ',<br/> > 'phonemes':'iyɨʉɯuɪʏʊeøɘəɵɤoɛœɜɞʌɔæɐaɶɑɒᵻʘɓǀɗǃʄǂɠǁʛpbtdʈɖcɟkɡqɢʔɴŋɲɳnɱmʙrʀⱱɾɽɸβfvθðszʃʒʂʐçʝxɣχʁħʕhɦɬɮʋɹɻjɰlɭʎʟˈˌːˑʍwɥʜʢʡɕʑɺɧɚ˞ɫ'<br/> > },<br/> ><br/> > 'reinit_layers': [], // give a list of layer names to restore from the given checkpoint. If not defined, it reloads all heuristically matching layers.<br/> ><br/> > // TRAINING<br/> > 'batch_size': 32, // Batch size for training. Lower values than 32 might cause hard to learn attention. It is overwritten by 'gradual_training'.<br/> > 'eval_batch_size':16, <br/> > 'r': 7, // Number of decoder frames to predict per iteration. Set the initial values if gradual training is enabled. <br/> > 'gradual_training': [[0, 7, 64], [20, 5, 64], [5000, 3, 32], [15000, 2, 32], [27000, 1, 16]],<br/> > 'loss_masking': true, // enable / disable loss masking against the sequence padding.<br/> ><br/> ><br/> > // OPTIMIZER<br/> > 'noam_schedule': false, // use noam warmup and lr schedule.<br/> > 'grad_clip': 1.0, // upper limit for gradients for clipping.<br/> > 'epochs': 10000, // total number of epochs to train.<br/> > 'lr': 0.0001, // Initial learning rate. If Noam decay is active, maximum learning rate.<br/> > 'wd': 0.000001, // Weight decay weight.<br/> > 'warmup_steps': 4000, // Noam decay steps to increase the learning rate from 0 to 'lr'<br/> > 'seq_len_norm': true, // Normalize eash sample loss with its length to alleviate imbalanced datasets. Use it if your dataset is small or has skewed distribution of sequence lengths.<br/> ><br/> > // TACOTRON PRENET<br/> > 'memory_size': -1, // ONLY TACOTRON - size of the memory queue used fro storing last decoder predictions for auto-regression. If < 0, memory queue is disabled and decoder only uses the last prediction frame.<br/> > 'prenet_type': 'original', // 'original' or 'bn'.<br/> > 'prenet_dropout': true, // enable/disable dropout at prenet.<br/> ><br/> > // ATTENTION<br/> > 'attention_type': 'original', // 'original' or 'graves'<br/> > 'attention_heads': 4, // number of attention heads (only for 'graves')<br/> > 'attention_norm': 'softmax', // softmax or sigmoid. Suggested to use softmax for Tacotron2 and sigmoid for Tacotron.<br/> > 'windowing': false, // Enables attention windowing. Used only in eval mode.<br/> > 'use_forward_attn': true, // if it uses forward attention. In general, it aligns faster.<br/> > 'forward_attn_mask': false, // Additional masking forcing monotonicity only in eval mode.<br/> > 'transition_agent': false, // enable/disable transition agent of forward attention.<br/> > 'location_attn': true, // enable_disable location sensitive attention. It is enabled for TACOTRON by default.<br/> > 'bidirectional_decoder': false, // use https://arxiv.org/abs/1907.09006. Use it, if attention does not work well with your dataset.<br/> ><br/> > // STOPNET<br/> > 'stopnet': true, // Train stopnet predicting the end of synthesis.<br/> > 'separate_stopnet': true, // Train stopnet seperately if 'stopnet==true'. It prevents stopnet loss to influence the rest of the model. It causes a better model, but it trains SLOWER.<br/> ><br/> > // DATA LOADING<br/> > 'text_cleaner': 'en_cleaners',<br/> > 'enable_eos_bos_chars': false, // enable/disable beginning of sentence and end of sentence chars.<br/> > 'num_loader_workers': 4, // number of training data loader processes. Don't set it too big. 4-8 are good values.<br/> > 'num_val_loader_workers': 4, // number of evaluation data loader processes.<br/> > 'batch_group_size': 0, //Number of batches to shuffle after bucketing.<br/> > 'min_seq_len': 6, // DATASET-RELATED: minimum text length to use in training<br/> > 'max_seq_len': 300, // DATASET-RELATED: maximum text length<br/> ><br/> > // PHONEMES<br/> > 'phoneme_cache_path': 'mozilla_us_phonemes_2_1', // phoneme computation is slow, therefore, it caches results in the given folder.<br/> > 'use_phonemes': false, // use phonemes instead of raw characters. It is suggested for better pronounciation.<br/> > 'phoneme_language': 'en-us', // depending on your target language, pick one from https://github.com/bootphon/phonemizer#languages<br/> ><br/> > // MULTI-SPEAKER and GST<br/> > 'use_speaker_embedding': false, // use speaker embedding to enable multi-speaker learning.<br/> > 'style_wav_for_test': null, // path to style wav file to be used in TacotronGST inference.<br/> > 'use_gst': false, // TACOTRON ONLY: use global style tokens<br/> ><br/> > }<br/> <br/> Thanks in advance!<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/stop-loss-spikes]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> maneeshkyadav<br/> <DATE> [April 6, 2020, 10:44pm]<br/> <br/> I had the same problem with 'master' and I am not getting it in 'dev'.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [April 7, 2020, 9:11am]<br/> <br/> try trimming silences in your dataset.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> rui<br/> <DATE> [April 7, 2020, 9:36am]<br/> <br/> Thanks. Running it now to see how it goes.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> rui<br/> <DATE> [April 7, 2020, 9:41am]<br/> <br/> You mean the silences at the beginning and end of the audio file,<br/> right? *slash* <br/> Are there any guidelines on how long these silences can be (at most)?<br/> I've checked some audio files from the LJ dataset using audacity, and it<br/> seems the silences can go up to .1 seconds.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [April 9, 2020, 10:43am]<br/> <br/> .1 is good enough. That should be fine<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> rui<br/> <DATE> [May 3, 2020, 8:39pm]<br/> <br/> Tried both suggestions, but the stop loss spikes keep happening. A small<br/> jump at r=2, and then a big jump r=1.<br/> <br/> What was the dev<br/> branch commit you tried? I've tried master branch commit **fab74dd** and<br/> dev branch **f7b1ca**<br/> <br/> I indeed found some clips in<br/> my dataset with silences at the beginning and at the end which were over<br/> .1seconds. After cleaning all those clips the problem still persists.<br/> Any other idea of what may be causing this?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> maneeshkyadav<br/> <DATE> [May 4, 2020, 3:28am]<br/> <br/> I cloned from dev branch commit:<br/> 2a15e391669f9073ba10ef7ff20bb54ec5246977<br/> <br/> Things seems to have been working ok across a bunch of runs. I can post<br/> graphs for successful runs if needed (spike graphs are in another<br/> thread).<br/> <br/> [Archived Post]<br/>
